src/minichain/storage.py

```
@@ -0,0 +1,280 @@
+            previous_hash=str(previous_hash),
+            merkle_root=str(merkle_root),
+            timestamp=int(timestamp),
+            difficulty_target=int(difficulty_target),
+            nonce=int(nonce),
+            block_height=int(height),
+        )
+        transaction_dicts = json.loads(str(transactions_json))
+        transactions = [Transaction(**tx) for tx in transaction_dicts]
+        block = Block(header=header, transactions=transactions)
+        if not block.has_valid_merkle_root():
+            raise StorageError("Corrupt block data: merkle_root mismatch")
```

Ruff (0.15.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)